### PR TITLE
Fix rule 18 to check non-fact concepts

### PR DESCRIPTION
**Pull request by sagesmith-wf**
#### Changes:

- Changed rule 18 to check deprecated concepts of non-facts too

Please review: @hefischer  @andrewperkins-wf

### CODE REVIEW

File: dqc_us_rules/dqc_us_0018.py [hunk 1]
```diff
@@ -244,14 +254,31 @@ def _fact_checkable(fact):
...
 
+def _catch_linkbase_deprecated_errors(val):
+    """
+    Check for unused concept relationships of standard taxonomy elements
+    and catches abstract deprecated concepts in linkbases
+
+    :param val:
+    :param ugtNamespace:
+    :return:
+    :rtype:
```
Comment (chrislococo-wf) [inline dqc_us_rules/dqc_us_0018.py]: SG: Incomplete Docstring

**Comment by mindeerickson-wf** [comment]
+1

**Comment by brianmyers-wf** [comment]
+1

**Comment by mindeerickson-wf** [comment]
+1

File: dqc_us_rules/dqc_us_0018.py [hunk 2]
```diff
@@ -262,20 +259,22 @@ def _fact_checkable(fact):
...
     :return: Returns deprecated concepts from the linkbase
-    :rtype: :class: '~arelle.ModelInstanceObject.ModelConcept'
+    :rtype: None
     """
     relationships = val.modelXbrl.relationshipSet(XbrlConst.parentChild)
     for rel in relationships.modelRelationships:
         for concept in (rel.fromModelObject, rel.toModelObject):
             if _deprecated_concept(val, concept):
-                yield concept
+                if not deprecated_concepts.get(concept.name):
```
Comment (brianmyers-wf) [inline dqc_us_rules/dqc_us_0018.py]: there's a concept.name attribute? I thought you had to grab the qname localname?

**Comment by brianmyers-wf** [comment]
Do unit tests not need to be tweaked with these changes?

**Comment by mindeerickson-wf** [comment]
+1

File: dqc_us_rules/dqc_us_0018.py [hunk 3]
```diff
@@ -276,18 +301,20 @@ def _fact_uses_deprecated_item(val, fact):
...
 
 
-def _catch_deprecated_errors(val):
+def _catch_deprecated_fact_errors(val, deprecated_concepts):
     """
     Checks to see if facts are using deprecated items
     :param val: ValidateXbrl to check for deprecated item
 
     :type val: :class:'~arelle.ValidateXbrl.ValidateXbrl'
     :return: Returns information needed to create error message
```
Comment (jasonleinbach-wf) [inline dqc_us_rules/dqc_us_0018.py]: Need to update the return docstring.

File: dqc_us_rules/dqc_us_0018.py [hunk 4]
```diff
@@ -244,14 +251,32 @@ def _fact_checkable(fact):
...
     )
 
 
+def _catch_linkbase_deprecated_errors(val, deprecated_concepts):
+    """
+    Check for unused concept relationships of standard taxonomy elements
+    and catches abstract deprecated concepts in linkbases
+
+    :param val: :class: '~arelle.ValdiateXbrl.ValidateXbrl'
+    :return: Returns deprecated concepts from the linkbase
```
Comment (jasonleinbach-wf) [inline dqc_us_rules/dqc_us_0018.py]: No return.

**Comment by jasonleinbach-wf** [comment]
+1

File: tests/unit_tests/test_dqc_us_0018.py [hunk 5]
```diff
@@ -113,3 +113,28 @@ def test_deprecated_member(self):
...
            )
         )
+
+    def test_catch_linkbase_deprecated_errors(self):
+        """
+        Tests for deprecated concepts in relationships
+        """
+        val = mock.Mock()
+        val.modelXbrl.namespaceDocs = {}
+        val.modelXbrl.namespaceDocs["http://fasb.org/us-gaap/2015-01-31"] = [0]
```
Comment (brianmyers-wf) [inline tests/unit_tests/test_dqc_us_0018.py]: these lines can be combined to:
```
val.modelXbrl.namespaceDocs = {"http://fasb.org/us-gaap/2015-01-31": [0]}
```

File: tests/unit_tests/test_dqc_us_0018.py [hunk 6]
```diff
@@ -113,3 +113,28 @@ def test_deprecated_member(self):
...
+        deprecated_concept = "PolicyChargesInsurance"
+        from_model = mock.Mock()
+        from_model.name = deprecated_concept
```
Comment (brianmyers-wf) [inline tests/unit_tests/test_dqc_us_0018.py]: These lines can be combined to:
```
from_model = mock.Mock(name=deprecated_concept)
to_model = mock.Mock(name=deprecated_concept)
```

File: tests/unit_tests/test_dqc_us_0018.py [hunk 7]
```diff
@@ -113,3 +113,28 @@ def test_deprecated_member(self):
...
+        to_model = mock.Mock()
+        to_model.name = deprecated_concept
+        val.usgaapDeprecations = [deprecated_concept]
+        relationship = mock.Mock(
+            fromModelObject=from_model,
+            toModelObject=to_model
+        )
+        relset = mock.Mock()
+        relset.modelRelationships = [relationship]
```
Comment (brianmyers-wf) [inline tests/unit_tests/test_dqc_us_0018.py]: These can be combined to:
```
relset = mock.Mock(modelRelationships=[relationship])
```

**Comment by brianmyers-wf** [comment]
pass complete

File: tests/unit_tests/test_dqc_us_0018.py (hunk 6 above)
Comment (sagesmith-wf) [inline tests/unit_tests/test_dqc_us_0018.py]: Does not work, I have already tried this

File: tests/unit_tests/test_dqc_us_0018.py (hunk 7 above)
Comment (sagesmith-wf) [inline tests/unit_tests/test_dqc_us_0018.py]: Does not work have already tried this

**Comment by brianmyers-wf** [comment]
+1

**Comment by mindeerickson-wf** [comment]
+1

**Comment by mindeerickson-wf** [comment]
Found a flake8 error:
(dqc) mindeerickson [dqc_us_rules] (XS-3122) $ flake8 dqc_us_rules/
dqc_us_rules/dqc_us_0018.py:277:80: E501 line too long (80 > 79 characters)

**Comment by mindeerickson-wf** [comment]
+1

**Comment by brianmyers-wf** [comment]
refresh +1

**Comment by mindeerickson-wf** [comment]
Testing notes can be found in the Jira ticket.

QA +1

**Comment by jasonleinbach-wf** [comment]
Reviewed and verified.  Merging.

**Pull request merged**